test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs

```diff
@@ -0,0 +1,280 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Cardano.Wallet.CoinSelection.LargestFirstSpec
+    ( spec
+    ) where
+
+import Prelude
+
+import Cardano.Wallet.CoinSelection
+    ( CoinSelection (..), CoinSelectionError (..), CoinSelectionOptions (..) )
+import Cardano.Wallet.CoinSelection.LargestFirst
+    ( largestFirst )
+import Cardano.Wallet.Primitive.Types
+    ( Address (..)
+    , Coin (..)
+    , Hash (..)
+    , TxIn (..)
+    , TxOut (..)
+    , UTxO (..)
+    , invariant
+    )
+import Control.DeepSeq
+    ( deepseq )
+import Control.Monad.IO.Class
+    ( liftIO )
+import Control.Monad.Trans.Except
+    ( runExceptT )
+import Data.List.NonEmpty
+    ( NonEmpty (..) )
+import Data.Word
+    ( Word64 )
+import Test.Hspec
+    ( Spec, describe, it, shouldBe )
+import Test.QuickCheck
+    ( Arbitrary (..)
+    , Property
+    , choose
+    , oneof
+    , property
+    , scale
+    , suchThat
+    , vectorOf
+    )
+import Test.QuickCheck.Monadic
+    ( monadicIO )
+
+import qualified Data.List as L
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Map.Strict as Map
+
+
+spec :: Spec
+spec = do
+    describe "Coin selection : LargestFirst algorithm" $ do
+        it "works as expected for zero-fee and fully covered by utxo coin selection"
+            (property propLargestFirstFullyCovered)
+        it "works as expected for zero-fee and not covered by utxo coin selection"
+            (property propLargestFirstNotCovered)
+        it "works as expected for zero-fee and fully covered by utxo coin selection when maximumNumberOfInputs is small"
+            (property propLargestFirstFullyCoveredSmallMaxInput)
+
+{-------------------------------------------------------------------------------
+                                Properties
+-------------------------------------------------------------------------------}
+
+propLargestFirstFullyCovered
+    :: CoveringCase
+    -> Property
+propLargestFirstFullyCovered (CoveringCase (utxo, txOuts)) =
+    monadicIO $ liftIO $ do
+        let check =
+                invariant "utxo must cover all transaction outputs"
+                (NE.length txOuts)
+                (\_ -> condCoinsCovering txOuts utxo)
```

Comment:
You could use a precondition for this using QuickCheck's `(==>)`

Reply:
:+1: